### ISSUE TITLE
Cosmos-1_0-Diffusion-7B-Text2World-BF16.gguf'

### ISSUE THREAD

**Issue by al-swaiti** [city96/ComfyUI-GGUF#204]
i successfuly convert Cosmos-1_0-Diffusion-7B-Text2World.safetensors to gguf format using convert.py , but now i stuck on convert it to q_8, q_4 etc using llama.cpp ! which tag i must use is there any patch i must apply ?

**Comment by clarin-ebtio800090** [comment]
Hello, we tried to solve the issue.

This is what we did:

Add specific instructions for quantizing the Cosmos-1_0-Diffusion-7B-Text2World model using llama.cpp, including the correct tag to use and patch application.

You can review changes in this commit: https://github.com/clarin-ebtio800090/city96-ComfyUI-GGUF-204/commit/70d8c28c4183febd1a92d1ddcd5d4c32aaf0dfda.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

Latta AI seeks to solve problems in open source projects as part of its mission to support developers around the world. Learn more about our mission at https://latta.ai/ourmission . If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by al-swaiti** [comment]
thanks ,,, i tried to convert the model but it give me error 

![Image](https://github.com/user-attachments/assets/23f61bf9-919f-4b25-960d-d486185607d3)

**Comment by al-swaiti** [comment]
`import os
import torch
import gguf  # This needs to be the llama.cpp one specifically!
import argparse
from tqdm import tqdm

from safetensors.torch import load_file

# Define constants
MAX_TENSOR_NAME_LENGTH = 255
QUANTIZATION_THRESHOLD = 256  # Adjust as needed
REARRANGE_THRESHOLD = 256  # Adjust as needed

# Define a basic ModelTemplate class
class ModelTemplate:
    # This is a base class for defining model architectures
    arch = "template"  # Placeholder architecture name
    keys_detect = []  # List of key tuples to detect architecture
    keys_banned = []  # List of banned keys that we don't want to allow
    shape_fix = False  # Flag for shape rearrangement
    
    # Add any default methods or properties for handling models here if needed

# Define custom model architecture based on your keys
class ModelCustom(ModelTemplate):
    arch = "custom"  # Custom architecture name
    keys_detect = [
        ("logvar.0.freqs", "logvar.0.phases", "logvar.1.weight"),
        ("net.affline_norm.weight",),
        ("net.blocks.block0.blocks.0.adaLN_modulation.1.weight", "net.blocks.block0.blocks.0.adaLN_modulation.2.weight"),
        ("net.blocks.block0.blocks.0.block.attn.to_k.0.weight", "net.blocks.block0.blocks.0.block.attn.to_k.1.weight"),
        ("net.blocks.block0.blocks.0.block.attn.to_out.0.weight", "net.blocks.block0.blocks.0.block.attn.to_q.0.weight"),
    ]
    keys_banned = []  # No banned keys for this model, but you can add them if needed

# Define Cosmos model architecture
class ModelCosmos(ModelTemplate):
    arch = "cosmos"
    keys_detect = [
        ("logvar.0.freqs", "logvar.0.phases", "logvar.1.weight"),
        ("net.affline_norm.weight",),
        ("net.blocks.block0.blocks.0.adaLN_modulation.1.weight", "net.blocks.block0.blocks.0.adaLN_modulation.2.weight"),
        ("net.blocks.block0.blocks.0.block.attn.to_k.0.weight", "net.blocks.block0.blocks.0.block.attn.to_k.1.weight"),
        ("net.blocks.block0.blocks.0.block.attn.to_out.0.weight", "net.blocks.block0.blocks.0.block.attn.to_q.0.weight"),
    ]
    keys_banned = []
    shape_fix = True

# List of architectures including custom ones
arch_list = [ModelCustom, ModelCosmos]  # Include other models as needed

def is_model_arch(model, state_dict):
    # check if model is correct
    matched = False
    invalid = False
    for match_list in model.keys_detect:
        if all(key in state_dict for key in match_list):
            matched = True
            invalid = any(key in state_dict for key in model.keys_banned)
            break
    assert not invalid, "Model architecture not allowed for conversion! (i.e. reference VS diffusers format)"
    return matched

def detect_arch(state_dict):
    model_arch = None
    for arch in arch_list:
        if is_model_arch(arch, state_dict):
            model_arch = arch
            break
    assert model_arch is not None, "Unknown model architecture!"
    return model_arch

def parse_args():
    parser = argparse.ArgumentParser(description="Generate F16 GGUF files from single UNET")
    parser.add_argument("--src", required=True, help="Source model ckpt file.")
    parser.add_argument("--dst", help="Output unet gguf file.")
    args = parser.parse_args()

    if not os.path.isfile(args.src):
        parser.error("No input provided!")

    return args

def load_state_dict(path):
    if any(path.endswith(x) for x in [".ckpt", ".pt", ".bin", ".pth"]):
        state_dict = torch.load(path, map_location="cpu", weights_only=True)
        state_dict = state_dict.get("model", state_dict)
    else:
        state_dict = load_file(path)

    # only keep unet with no prefix!
    prefix = None
    for pfx in ["model.diffusion_model.", "model."]:
        if any([x.startswith(pfx) for x in state_dict.keys()]):
            prefix = pfx
            break

    sd = {}
    for k, v in state_dict.items():
        if prefix and prefix not in k:
            continue
        if prefix:
            k = k.replace(prefix, "")
        sd[k] = v

    return sd

def load_model(path):
    state_dict = load_state_dict(path)
    model_arch = detect_arch(state_dict)
    print(f"* Architecture detected from input: {model_arch.arch}")
    writer = gguf.GGUFWriter(path=None, arch=model_arch.arch)
    return (writer, state_dict, model_arch)

def handle_tensors(args, writer, state_dict, model_arch):
    name_lengths = tuple(sorted(
        ((key, len(key)) for key in state_dict.keys()),
        key=lambda item: item[1],
        reverse=True,
    ))
    if not name_lengths:
        return
    max_name_len = name_lengths[0][1]
    if max_name_len > MAX_TENSOR_NAME_LENGTH:
        bad_list = ", ".join(f"{key!r} ({namelen})" for key, namelen in name_lengths if namelen > MAX_TENSOR_NAME_LENGTH)
        raise ValueError(f"Can only handle tensor names up to {MAX_TENSOR_NAME_LENGTH} characters. Tensors exceeding the limit: {bad_list}")
    
    for key, data in tqdm(state_dict.items()):
        old_dtype = data.dtype

        if data.dtype == torch.bfloat16:
            data = data.to(torch.float32).numpy()
        elif data.dtype in [getattr(torch, "float8_e4m3fn", "_invalid"), getattr(torch, "float8_e5m2", "_invalid")]:
            data = data.to(torch.float16).numpy()
        else:
            data = data.numpy()

        n_dims = len(data.shape)
        data_shape = data.shape
        data_qtype = getattr(gguf.GGMLQuantizationType, "BF16" if old_dtype == torch.bfloat16 else "F16")

        n_params = 1
        for dim_size in data_shape:
            n_params *= dim_size

        blacklist = {
            "time_embedding.", "add_embedding.", "time_in.", "txt_in.", 
            "vector_in.", "img_in.", "guidance_in.", "final_layer.",
        }

        if old_dtype in (torch.float32, torch.bfloat16):
            if n_dims == 1:
                data_qtype = gguf.GGMLQuantizationType.F32
            elif n_params <= QUANTIZATION_THRESHOLD:
                data_qtype = gguf.GGMLQuantizationType.F32
            elif ".weight" in key and any(x in key for x in blacklist):
                data_qtype = gguf.GGMLQuantizationType.F32

        if (model_arch.shape_fix and n_dims > 1 and n_params >= REARRANGE_THRESHOLD 
            and (n_params / 256).is_integer() and not (data.shape[-1] / 256).is_integer()):
            orig_shape = data.shape
            data = data.reshape(n_params // 256, 256)
            writer.add_array(f"comfy.gguf.orig_shape.{key}", tuple(int(dim) for dim in orig_shape))

        try:
            data = gguf.quants.quantize(data, data_qtype)
        except (AttributeError, gguf.QuantError) as e:
            tqdm.write(f"falling back to F16: {e}")
            data_qtype = gguf.GGMLQuantizationType.F16
            data = gguf.quants.quantize(data, data_qtype)

        new_name = key

        shape_str = f"{{{', '.join(str(n) for n in reversed(data.shape))}}}"
        tqdm.write(f"{f'%-{max_name_len + 4}s' % f'{new_name}'} {old_dtype} --> {data_qtype.name}, shape = {shape_str}")

        writer.add_tensor(new_name, data, raw_dtype=data_qtype)

if __name__ == "__main__":
    args = parse_args()
    path = args.src
    writer, state_dict, model_arch = load_model(path)

    writer.add_quantization_version(gguf.GGML_QUANT_VERSION)
    if next(iter(state_dict.values())).dtype == torch.bfloat16:
        out_path = f"{os.path.splitext(path)[0]}-BF16.gguf"
        writer.add_file_type(gguf.LlamaFileType.MOSTLY_BF16)
    else:
        out_path = f"{os.path.splitext(path)[0]}-F16.gguf"
        writer.add_file_type(gguf.LlamaFileType.MOSTLY_F16)

    out_path = args.dst or out_path
    if os.path.isfile(out_path):
        input("Output exists. Press Enter to continue or Ctrl+C to abort!")

    handle_tensors(args, writer, state_dict, model_arch)
    writer.write_header_to_file(path=out_path)
    writer.write_kv_data_to_file()
    writer.write_tensors_to_file(progress=True)
    writer.close()`
this is convert.py code ,, i used but now i dont know which tag of llama.cpp  i must used to quantize the model nor how to make the patch
